File: ios/LocalPods/ScannerModule/Core/Detection/SegmentationHelper.swift
import Foundation
import CoreGraphics
import Accelerate

// MARK: - SegmentationHelper

/// Helper class to reconstruct segmentation masks from YOLOv2.6 Seg model outputs.
///
/// Model output:
///   Output[0]: [1, 300, 38] → 300 detections × [x1, y1, x2, y2, conf, classId, 32 coeffs]
///   Output[1]: [1, 160, 160, 32] → 32 prototype masks, each 160×160
///
/// Mask reconstruction:
///   mask[y, x] = sigmoid( Σ( coeff[i] × proto_mask[i, y, x] ) )  for i = 0..31
///
/// Mirrors Android SegmentationHelper.kt exactly.
enum SegmentationHelper {

    // MARK: - Constants

    /// Prototype mask dimension (160×160).
    static let protoSize = 160

    /// Number of prototype masks (mask channels).
    static let numProtos = 32

    // MARK: - Mask Reconstruction

    /// Reconstruct a 2D mask from coefficients and prototype masks.
    ///
    /// - Parameters:
    ///   - coeffs: 32 mask coefficients from output[0][i][6:38]
    ///   - protos: 32 prototype masks [32][160][160] from output[1]
    /// - Returns: 2D FloatArray [160][160] with sigmoid values [0, 1]
    ///
    /// Formula: mask[y, x] = sigmoid( Σ( coeffs[i] × protos[i][y][x] ) )
    static func reconstructMask(coeffs: [Float], protos: [[Float]]) -> [[Float]] {
        let height = protoSize
        let width = protoSize

        // Validate inputs
        guard coeffs.count >= numProtos else {
            print("[SegmentationHelper] ❌ Invalid coeffs count: \(coeffs.count), expected \(numProtos)")
            return [[Float]](repeating: [Float](repeating: 0, count: width), count: height)
        }
        guard protos.count >= numProtos else {
            print("[SegmentationHelper] ❌ Invalid protos count: \(protos.count), expected \(numProtos)")
            return [[Float]](repeating: [Float](repeating: 0, count: width), count: height)
        }

        // Initialize empty mask
        var mask = [[Float]](repeating: [Float](repeating: 0, count: width), count: height)

        // Weighted sum: Σ( coeff[i] × proto[i] )
        for y in 0..<height {
            for x in 0..<width {
                var sum: Float = 0
                for i in 0..<numProtos {
                    // protos layout: protos[channel][height][width]
                    guard i < protos.count else { continue }
                    let protoChannel = protos[i]
                    let idx = y * width + x
                    guard idx < protoChannel.count else {
                        print("[SegmentationHelper] ❌ Proto channel[\(i)] size mismatch: \(protoChannel.count), expected \(width * height)")
                        continue
                    }
                    sum += coeffs[i] * protoChannel[idx]
                }
                // Sigmoid activation
                mask[y][x] = sigmoid(sum)
            }
        }

        #if DEBUG
        let minVal = mask.flatMap { $0 }.min() ?? 0
        let maxVal = mask.flatMap { $0 }.max() ?? 0
        print("[SegmentationHelper] ✅ Mask reconstructed: \(width)×\(height), range=[\(minVal), \(maxVal)]")
        #endif

        return mask
    }

    /// Reconstruct mask with bounding box constraint in proto space (KEY FIX from Python).
    /// This improves mask quality by only computing sigmoid within the detection box region.
    /// Pixels outside the box are set to 0 (will be removed as background).
    ///
    /// - Parameters:
    ///   - coeffs: 32 mask coefficients
    ///   - protos: 32 prototype masks [32][160][160]
    ///   - boxNormalized: Bounding box in normalized coords [x1, y1, x2, y2] in range [0-1]
    ///   - modelSize: Model input size (default 640)
    /// - Returns: Mask [160][160] with sigmoid values [0, 1] inside box, 0 outside
    static func reconstructMaskWithBoxConstraint(
        coeffs: [Float],
        protos: [[Float]],
        boxNormalized: [Float],
        modelSize: Int = ScannerConfig.modelInputSize
    ) -> [[Float]] {
        let height = protoSize
        let width = protoSize

        guard coeffs.count >= numProtos, protos.count >= numProtos else {
            return [[Float]](repeating: [Float](repeating: 0, count: width), count: height)
        }

        // Convert normalized box [0-1] to model space [0-640]
        let x1_model = boxNormalized[0] * Float(modelSize)
        let y1_model = boxNormalized[1] * Float(modelSize)
        let x2_model = boxNormalized[2] * Float(modelSize)
        let y2_model = boxNormalized[3] * Float(modelSize)

        // Scale box from model space (640) to proto space (160)
        let scale = Float(protoSize) / Float(modelSize)
        let px1 = Int(max(0, min(Float(width), x1_model * scale)))
        let py1 = Int(max(0, min(Float(height), y1_model * scale)))
        let px2 = Int(max(0, min(Float(width), x2_model * scale)))
        let py2 = Int(max(0, min(Float(height), y2_model * scale)))

        // Initialize mask (all zeros - background will be removed)
        var mask = [[Float]](repeating: [Float](repeating: 0, count: width), count: height)

        // Compute mask ONLY within bounding box region in proto space
        // This matches Python: mask_cropped[py1:py2, px1:px2] = mask_low[py1:py2, px1:px2]
        for y in py1..<py2 {
            for x in px1..<px2 {
                var sum: Float = 0
                for i in 0..<numProtos {
                    guard i < protos.count else { continue }
                    let protoChannel = protos[i]
                    let idx = y * width + x
                    guard idx < protoChannel.count else { continue }
                    sum += coeffs[i] * protoChannel[idx]
                }
                // Apply sigmoid only within box region
                mask[y][x] = sigmoid(sum)
            }
        }

        #if DEBUG
        let boxArea = (px2 - px1) * (py2 - py1)
        print("[SegmentationHelper] ✅ Mask reconstructed with box constraint: proto_box=[\(px1),\(py1),\(px2),\(py2)], area=\(boxArea)")
        #endif

        return mask
    }

    /// Reconstruct mask from interleaved proto data (flat array [32 × 160 × 160]).
    /// Use this when protos come as a flat Float array from TFLite output.
    static func reconstructMask(coeffs: [Float], protoFlat: Data) -> [[Float]] {
        let height = protoSize
        let width = protoSize

        let floatCount = protoFlat.count / MemoryLayout<Float>.size
        guard let floatPtr = protoFlat.withUnsafeBytes({ $0.bindMemory(to: Float.self).baseAddress }) else {
            return [[Float]](repeating: [Float](repeating: 0, count: width), count: height)
        }

        var mask = [[Float]](repeating: [Float](repeating: 0, count: width), count: height)

        for y in 0..<height {
            for x in 0..<width {
                var sum: Float = 0
                for i in 0..<numProtos {
                    // protoFlat layout: [ch0_h0_w0, ch0_h0_w1, ..., ch0_h160_w159, ch1_h0_w0, ...]
                    let offset = i * protoSize * protoSize + y * protoSize + x
                    if offset < floatCount {
                        sum += coeffs[i] * floatPtr[offset]
                    }
                }
                mask[y][x] = sigmoid(sum)
            }
        }

        #if DEBUG
        let flat = mask.flatMap { $0 }
        let minVal = flat.min() ?? 0
        let maxVal = flat.max() ?? 0
        print("[SegmentationHelper] ✅ Mask reconstructed (flat): \(width)×\(height), range=[\(minVal), \(maxVal)]")
        #endif

        return mask
    }

    // MARK: - Mask Resizing

    /// Resize mask from 160×160 to the target dimensions using bilinear interpolation.
    ///
    /// - Parameters:
    ///   - mask: Source mask 160×160
    ///   - width: Target width (pixels)
    ///   - height: Target height (pixels)
    /// - Returns: Resized mask [height][width]
    static func resizeMaskToBox(_ mask: [[Float]], width: Int, height: Int) -> [[Float]] {
        guard width > 0, height > 0 else { return mask }

        let srcHeight = mask.count
        let srcWidth = mask.isEmpty ? 0 : mask[0].count
        guard srcWidth > 0, srcHeight > 0 else { return mask }

        var resized = [[Float]](repeating: [Float](repeating: 0, count: width), count: height)

        for y in 0..<height {
            for x in 0..<width {
                // Map to source coordinates
                let srcX = Float(x) / Float(width) * Float(srcWidth - 1)
                let srcY = Float(y) / Float(height) * Float(srcHeight - 1)

                let x0 = max(0, min(srcWidth - 2, Int(srcX)))
                let y0 = max(0, min(srcHeight - 2, Int(srcY)))
                let x1 = x0 + 1
                let y1 = y0 + 1

                let xFrac = srcX - Float(x0)
                let yFrac = srcY - Float(y0)

                // Bilinear interpolation
                let v00 = mask[y0][x0]
                let v01 = mask[y0][x1]
                let v10 = mask[y1][x0]
                let v11 = mask[y1][x1]

                let value = v00 * (1 - xFrac) * (1 - yFrac)
                          + v01 * xFrac       * (1 - yFrac)
                          + v10 * (1 - xFrac) * yFrac
                          + v11 * xFrac       * yFrac

                resized[y][x] = value
            }
        }

        #if DEBUG
        print("[SegmentationHelper] 📐 Mask resized: \(srcWidth)×\(srcHeight) → \(width)×\(height)")
        #endif

        return resized
    }

    // MARK: - Mask Application

    /// Apply a mask to a cropped image — background becomes black/transparent.
    ///
    /// - Parameters:
    ///   - image: Source image (from crop)
    ///   - mask: Resized mask matching image dimensions, values [0-1]
    ///   - threshold: Sigmoid threshold [0-1]. Default 0.5.
    ///     pixels with mask > threshold → kept; else → transparent/black.
    /// - Returns: New CGImage with masked background.
    static func applyMaskToBitmap(_ image: CGImage, mask: [[Float]], threshold: Float = ScannerConfig.maskThreshold) -> CGImage? {
        let width = image.width
        let height = image.height

        let maskHeight = mask.count
        let maskWidth = mask.isEmpty ? 0 : mask[0].count

        guard maskWidth > 0, maskHeight > 0 else { return image }

        // Read source pixels
        guard let sourceData = image.dataProvider?.data,
              let sourcePtr = CFDataGetBytePtr(sourceData) else {
            print("[SegmentationHelper] ❌ Failed to get image data")
            return nil
        }

        let bytesPerPixel = image.bitsPerPixel / 8
        let bytesPerRow = image.bytesPerRow

        // Create output buffer (ARGB)
        var outputPixels = [UInt8](repeating: 0, count: width * height * 4)

        for y in 0..<height {
            for x in 0..<width {
                let pixelOffset = y * bytesPerRow + x * bytesPerPixel

                // Map to mask coordinates
                let maskX = max(0, min(maskWidth - 1, x * maskWidth / max(width, 1)))
                let maskY = max(0, min(maskHeight - 1, y * maskHeight / max(height, 1)))

                let maskValue = mask[maskY][maskX]

                let outOffset = (y * width + x) * 4
                if maskValue > threshold {
                    // Keep the original pixel (RGBA)
                    outputPixels[outOffset + 0] = sourcePtr[pixelOffset + 0]
                    outputPixels[outOffset + 1] = sourcePtr[pixelOffset + 1]
                    outputPixels[outOffset + 2] = sourcePtr[pixelOffset + 2]
                    outputPixels[outOffset + 3] = 255 // Force opaque
                } else {
                    // Solid black background (to match Python script and avoid JPEG alpha-to-white issue)
                    outputPixels[outOffset + 0] = 0
                    outputPixels[outOffset + 1] = 0
                    outputPixels[outOffset + 2] = 0
                    outputPixels[outOffset + 3] = 255
                }
            }
        }

        return createCGImage(from: outputPixels, width: width, height: height)
    }

    /// Apply mask with feathered (soft) edges.
    ///
    /// - Parameters:
    ///   - image: Source cropped image
    ///   - mask: Resized mask matching image dimensions
    ///   - innerThreshold: Inner threshold (vùng chắc chắn giữ lại, default 0.6)
    ///   - outerThreshold: Outer threshold (vùng chắc chắn bỏ, default 0.3)
    /// - Returns: CGImage with feathered mask edges.
    static func applyMaskWithFeather(
        _ image: CGImage,
        mask: [[Float]],
        innerThreshold: Float = ScannerConfig.maskInnerThreshold,
        outerThreshold: Float = ScannerConfig.maskOuterThreshold
    ) -> CGImage? {
        let width = image.width
        let height = image.height

        let maskHeight = mask.count
        let maskWidth = mask.isEmpty ? 0 : mask[0].count

        guard maskWidth > 0, maskHeight > 0 else { return image }

        guard let sourceData = image.dataProvider?.data,
              let sourcePtr = CFDataGetBytePtr(sourceData) else {
            return nil
        }

        let bytesPerPixel = image.bitsPerPixel / 8
        let bytesPerRow = image.bytesPerRow

        var outputPixels = [UInt8](repeating: 0, count: width * height * 4)

        for y in 0..<height {
            for x in 0..<width {
                let pixelOffset = y * bytesPerRow + x * bytesPerPixel

                let maskX = max(0, min(maskWidth - 1, x * maskWidth / max(width, 1)))
                let maskY = max(0, min(maskHeight - 1, y * maskHeight / max(height, 1)))
                let maskValue = mask[maskY][maskX]

                let r = sourcePtr[pixelOffset + 0]
                let g = sourcePtr[pixelOffset + 1]
                let b = sourcePtr[pixelOffset + 2]
                let a = sourcePtr[pixelOffset + 3]

                let outOffset = (y * width + x) * 4

                switch maskValue {
                case _ where maskValue >= innerThreshold:
                    // Inner: keep full pixel
                    outputPixels[outOffset + 0] = r
                    outputPixels[outOffset + 1] = g
                    outputPixels[outOffset + 2] = b
                    outputPixels[outOffset + 3] = a

                case _ where maskValue <= outerThreshold:
                    // Outer: Solid Black (Opaque)
                    // We must set alpha = 255, otherwise JPEG compression evaluates alpha=0 as pure White!
                    outputPixels[outOffset + 0] = 0   // R
                    outputPixels[outOffset + 1] = 0   // G
                    outputPixels[outOffset + 2] = 0   // B
                    outputPixels[outOffset + 3] = 255 // A

                default:
                    // Transition zone: blend alpha
                    let alpha = (maskValue - outerThreshold) / (innerThreshold - outerThreshold)
                    let clampedAlpha = max(0, min(1, alpha))
                    let newAlpha = UInt8(clampedAlpha * 255)

                    outputPixels[outOffset + 0] = r
                    outputPixels[outOffset + 1] = g
                    outputPixels[outOffset + 2] = b
                    outputPixels[outOffset + 3] = newAlpha
                }
            }
        }

        #if DEBUG
        print("[SegmentationHelper] 🎭 Feathered mask applied: \(width)×\(height)")
        #endif

        return createCGImage(from: outputPixels, width: width, height: height)
    }

    // MARK: - Mask Validation

    /// Check if a mask is valid (not all-zero and no NaN/Inf values).
    static func isValidMask(_ mask: [[Float]]) -> Bool {
        guard !mask.isEmpty, !mask[0].isEmpty else { return false }

        var sum: Float = 0
        var count = 0

        for row in mask {
            for v in row {
                if v.isNaN || v.isInfinite { return false }
                sum += v
                count += 1
            }
        }

        // Mean must be > 0.01 for a valid mask
        return count > 0 && (sum / Float(count)) > 0.01
    }

    // MARK: - Debug

    /// Print mask statistics for debugging.
    static func debugMask(_ mask: [[Float]]) {
        guard !mask.isEmpty else {
            print("[SegmentationHelper] 📊 Mask: empty")
            return
        }

        let flat = mask.flatMap { $0 }
        let minVal = flat.min() ?? 0
        let maxVal = flat.max() ?? 0
        let sum = flat.reduce(0, +)
        let mean = sum / Float(flat.count)

        print("[SegmentationHelper] 📊 Mask stats: \(mask[0].count)×\(mask.count), min=\(minVal), max=\(maxVal), mean=\(mean)")
    }

    // MARK: - Private Helpers

    /// Sigmoid activation with numerical stability.
    /// Mirrors Android sigmoid() function.
    private static func sigmoid(_ x: Float) -> Float {
        if x >= 0 {
            let e = exp(-Double(x))
            return Float(1.0 / (1.0 + e))
        } else {
            let e = exp(Double(x))
            return Float(e / (1.0 + e))
        }
    }

    /// Create a CGImage from raw pixel data.
    private static func createCGImage(from pixels: [UInt8], width: Int, height: Int) -> CGImage? {
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        let bitmapInfo = CGBitmapInfo(rawValue: CGImageAlphaInfo.premultipliedLast.rawValue)

        guard let provider = CGDataProvider(data: Data(pixels) as CFData) else {
            return nil
        }

        return CGImage(
            width: width,
            height: height,
            bitsPerComponent: 8,
            bitsPerPixel: 32,
            bytesPerRow: width * 4,
            space: colorSpace,
            bitmapInfo: bitmapInfo,
            provider: provider,
            decode: nil,
            shouldInterpolate: false,
            intent: .defaultIntent
        )
    }
}

File: android/orilifesdk/src/main/java/com/mvp/orilife/detection/SegmentationHelper.kt
package com.mvp.orilife.detection

import android.graphics.Bitmap
import android.graphics.Color
import android.util.Log
import com.mvp.orilife.Config
import kotlin.math.exp

/**
 * Helper class để reconstruct mask từ YOLOv2.6 Seg model outputs.
 *
 * Model output:
 *   Output[0]: [1, 300, 38] → 300 detections × [cx, cy, w, h, conf, classId, 32 coeffs]
 *   Output[1]: [1, 160, 160, 32] → 32 prototype masks, mỗi 160×160
 *
 * Mask reconstruction:
 *   mask[y, x] = sigmoid( Σ( coeff[i] × proto_mask[i, y, x] ) )  với i = 0..31
 */
object SegmentationHelper {

    private const val TAG = "SegmentationHelper"

    // Kích thước prototype mask từ model
    const val PROTO_SIZE = 160

    // Số prototype masks (mask channels)
    const val NUM_PROTOS = 32

    /**
     * Reconstruct mask từ coefficients và prototype masks
     *
     * @param coeffs 32 mask coefficients từ output[0][i][6:38]
     * @param protos 32 prototype masks, mỗi 160×160 từ output[1]
     * @return 2D FloatArray [160][160] với giá trị sigmoid [0, 1]
     */
    fun reconstructMask(
        coeffs: FloatArray,
        protos: Array<Array<FloatArray>>
    ): Array<Array<Float>> {
        val height = PROTO_SIZE
        val width = PROTO_SIZE

        // Validate protos shape: should be [32][160][160] = [NUM_PROTOS][PROTO_SIZE][PROTO_SIZE]
        if (protos.size != NUM_PROTOS) {
            Log.e(TAG, "⚠️ Proto shape mismatch: expected ${NUM_PROTOS} protos, got ${protos.size}")
        }
        if (protos.isNotEmpty() && protos[0].size != PROTO_SIZE) {
            Log.e(TAG, "⚠️ Proto H mismatch: expected ${PROTO_SIZE}, got ${protos[0].size}")
        }
        if (protos.isNotEmpty() && protos[0].isNotEmpty() && protos[0][0].size != PROTO_SIZE) {
            Log.e(TAG, "⚠️ Proto W mismatch: expected ${PROTO_SIZE}, got ${protos[0][0].size}")
        }

        // Khởi tạo mask rỗng
        val mask = Array(height) { Array(width) { 0f } }

        // Tính tổng có trọng số: Σ( coeff[i] × proto[i] )
        var minVal = Float.MAX_VALUE
        var maxVal = Float.MIN_VALUE
        var sumVal = 0f
        for (y in 0 until height) {
            for (x in 0 until width) {
                var sum = 0f
                for (i in 0 until minOf(NUM_PROTOS, protos.size)) {
                    val protoH = protos[i].size
                    val protoW = if (protoH > 0) protos[i][0].size else 0
                    if (y < protoH && x < protoW) {
                        sum += coeffs[i] * protos[i][y][x]
                    }
                }
                mask[y][x] = sigmoid(sum)
                if (mask[y][x] < minVal) minVal = mask[y][x]
                if (mask[y][x] > maxVal) maxVal = mask[y][x]
                sumVal += mask[y][x]
            }
        }

        val meanVal = sumVal / (height * width)
        val coeffMax = coeffs.maxOrNull() ?: 0f
        Log.d(TAG, "✅ Mask reconstructed: ${width}x${height}, range=[${"%.4f".format(minVal)}, ${"%.4f".format(maxVal)}], mean=${"%.4f".format(meanVal)}, coeffMax=${"%.4f".format(coeffMax)}")
        return mask
    }

    /**
     * Sigmoid function với numerical stability
     */
    private fun sigmoid(x: Float): Float {
        return when {
            x >= 0 -> {
                val e = exp(-x.toDouble())
                (1.0 / (1.0 + e)).toFloat()
            }
            else -> {
                val e = exp(x.toDouble())
                (e / (1.0 + e)).toFloat()
            }
        }
    }

    /**
     * Resize mask 160×160 → kích thước của bounding box (trong ảnh crop)
     *
     * @param mask Mask gốc 160×160
     * @param boxWidth Chiều rộng box đích (pixels)
     * @param boxHeight Chiều cao box đích (pixels)
     * @return Mask đã resize
     */
    fun resizeMaskToBox(
        mask: Array<Array<Float>>,
        boxWidth: Int,
        boxHeight: Int
    ): Array<Array<Float>> {
        if (boxWidth <= 0 || boxHeight <= 0) return mask

        val srcHeight = mask.size
        val srcWidth = if (mask.isNotEmpty()) mask[0].size else 0

        val resized = Array(boxHeight) { y ->
            Array(boxWidth) { x ->
                // Bilinear interpolation
                val srcX = (x.toFloat() / boxWidth * (srcWidth - 1)).coerceIn(0f, (srcWidth - 1).toFloat())
                val srcY = (y.toFloat() / boxHeight * (srcHeight - 1)).coerceIn(0f, (srcHeight - 1).toFloat())

                val x0 = srcX.toInt().coerceIn(0, srcWidth - 2)
                val y0 = srcY.toInt().coerceIn(0, srcHeight - 2)
                val x1 = x0 + 1
                val y1 = y0 + 1

                val xFrac = srcX - x0
                val yFrac = srcY - y0

                // Bilinear interpolation
                val v00 = mask[y0][x0]
                val v01 = mask[y0][x1]
                val v10 = mask[y1][x0]
                val v11 = mask[y1][x1]

                (v00 * (1 - xFrac) * (1 - yFrac) +
                        v01 * xFrac * (1 - yFrac) +
                        v10 * (1 - xFrac) * yFrac +
                        v11 * xFrac * yFrac)
            }
        }

        Log.d(TAG, "📐 Mask resized: ${srcWidth}x${srcHeight} → ${boxWidth}x${boxHeight}")
        return resized
    }

    /**
     * Áp mask vào bitmap đã crop - biến vùng ngoài mask thành đen
     *
     * @param croppedBitmap Bitmap đã crop (từ ImageCropper)
     * @param mask Mask đã resize theo kích thước crop
     * @param threshold Ngưỡng sigmoid [0-1], mặc định 0.5
     * @return Bitmap mới với background đen, chỉ vùng mask giữ nguyên
     */
    fun applyMaskToBitmap(
        croppedBitmap: Bitmap,
        mask: Array<Array<Float>>,
        threshold: Float = Config.MASK_THRESHOLD
    ): Bitmap {
        val width = croppedBitmap.width
        val height = croppedBitmap.height

        // Tạo bitmap mới để không modify bitmap gốc
        val result = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)

        val maskHeight = mask.size
        val maskWidth = if (mask.isNotEmpty()) mask[0].size else 0

        val pixels = IntArray(width * height)
        croppedBitmap.getPixels(pixels, 0, width, 0, 0, width, height)

        val resultPixels = IntArray(width * height)

        for (y in 0 until height) {
            for (x in 0 until width) {
                val idx = y * width + x
                val pixel = pixels[idx]

                // Map sang mask coordinates
                val maskX = (x.toFloat() / width * maskWidth).toInt().coerceIn(0, maskWidth - 1)
                val maskY = (y.toFloat() / height * maskHeight).toInt().coerceIn(0, maskHeight - 1)

                val maskValue = mask[maskY][maskX]

                if (maskValue > threshold) {
                    // Giữ nguyên pixel
                    resultPixels[idx] = pixel
                } else {
                    // Background → trong suốt (transparent)
                    resultPixels[idx] = Color.TRANSPARENT
                }
            }
        }

        result.setPixels(resultPixels, 0, width, 0, 0, width, height)

        // Tính % vùng giữ lại
        val keptPixels = resultPixels.count { it != Color.BLACK }
        val keptPercent = keptPixels * 100 / (width * height)
        Log.d(TAG, "🎭 Mask applied: ${width}x${height}, kept=$keptPercent% (threshold=$threshold)")

        return result
    }

    /**
     * Áp mask với độ mềm ở edges ( feathered mask )
     * Edge pixels sẽ có alpha transition thay vì hard cutoff
     *
     * @param croppedBitmap Bitmap đã crop
     * @param mask Mask đã resize
     * @param innerThreshold Ngưỡng trong (vùng chắc chắn giữ lại)
     * @param outerThreshold Ngưỡng ngoài (vùng chắc chắn bỏ)
     * @return Bitmap mới
     */
    fun applyMaskWithFeather(
        croppedBitmap: Bitmap,
        mask: Array<Array<Float>>,
        innerThreshold: Float = 0.6f,
        outerThreshold: Float = 0.3f
    ): Bitmap {
        val width = croppedBitmap.width
        val height = croppedBitmap.height

        val result = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)

        val maskHeight = mask.size
        val maskWidth = if (mask.isNotEmpty()) mask[0].size else 0

        val pixels = IntArray(width * height)
        croppedBitmap.getPixels(pixels, 0, width, 0, 0, width, height)

        val resultPixels = IntArray(width * height)

        for (y in 0 until height) {
            for (x in 0 until width) {
                val idx = y * width + x
                val pixel = pixels[idx]
                val r = (pixel shr 16) and 0xFF
                val g = (pixel shr 8) and 0xFF
                val b = pixel and 0xFF

                val maskX = (x.toFloat() / width * maskWidth).toInt().coerceIn(0, maskWidth - 1)
                val maskY = (y.toFloat() / height * maskHeight).toInt().coerceIn(0, maskHeight - 1)

                val maskValue = mask[maskY][maskX]

                when {
                    maskValue >= innerThreshold -> {
                        // Vùng trong - giữ nguyên
                        resultPixels[idx] = pixel
                    }
                    maskValue <= outerThreshold -> {
                        // Vùng ngoài - transparent
                        resultPixels[idx] = Color.TRANSPARENT
                    }
                    else -> {
                        // Vùng transition - blend với transparent (giảm alpha)
                        val alpha = ((maskValue - outerThreshold) / (innerThreshold - outerThreshold)).coerceIn(0f, 1f)
                        val newAlpha = (alpha * 255).toInt().coerceIn(0, 255)
                        resultPixels[idx] = (newAlpha shl 24) or (r shl 16) or (g shl 8) or b
                    }
                }
            }
        }

        result.setPixels(resultPixels, 0, width, 0, 0, width, height)
        Log.d(TAG, "🎭 Feathered mask applied: ${width}x${height}")
        return result
    }

    /**
     * Kiểm tra mask có valid không (không phải toàn zero hoặc NaN)
     */
    fun isValidMask(mask: Array<Array<Float>>): Boolean {
        if (mask.isEmpty() || mask[0].isEmpty()) return false

        var sum = 0f
        var count = 0
        for (row in mask) {
            for (v in row) {
                if (v.isNaN() || v.isInfinite()) return false
                sum += v
                count++
            }
        }

        // Trung bình phải > 0.01 mới là mask có nghĩa
        return count > 0 && (sum / count) > 0.01f
    }

    /**
     * Debug: in mask stats
     */
    fun debugMask(mask: Array<Array<Float>>) {
        val height = mask.size
        val width = if (mask.isNotEmpty()) mask[0].size else 0

        var sum = 0f
        var min = Float.MAX_VALUE
        var max = Float.MIN_VALUE

        for (row in mask) {
            for (v in row) {
                sum += v
                if (v < min) min = v
                if (v > max) max = v
            }
        }

        val count = width * height
        Log.d(TAG, "📊 Mask stats: ${width}x${height}, min=${"%.4f".format(min)}, max=${"%.4f".format(max)}, mean=${"%.4f".format(sum / count)}")
    }
}
